packages/testing/src/chai-retry-plugin/chai-retry-plugin.ts

```diff
@@ -0,0 +1,187 @@
+import Chai, { AssertionError } from 'chai';
+import { timeout as addTimeoutToPromise, sleep } from 'promise-assist';
+
+import type {
+    AssertionMethod,
+    FunctionToRetry,
+    AssertionStackItem,
+    RetryAndAssertProps,
+    RetryOptions,
+    PromiseLikeAssertion,
+    AssertionPropertyKeys,
+} from './types';
+
+const { Assertion, expect } = Chai;
+
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Chai {
+        interface Assertion {
+            /**
+             * Allows to retry the function passed to `expect` and assert the result until retries ended or timeout exceeded
+             * @param options Settings for retry logic:
+             * - `timeout`: The maximum duration in milliseconds to wait before failing the retry operation.
+             * - `retries`: The number of times to retry the function before failing.
+             * - `delay`: The delay in milliseconds between retries.
+             */
+            retry(options?: RetryOptions): PromiseLikeAssertion;
+        }
+    }
+}
+
+const retryFunctionAndAssertResult = async ({ functionToRetry, options, assertionStack }: RetryAndAssertProps) => {
+    const { retries, delay } = options;
+    let retriesCount = 0;
+
+    while (retriesCount < retries) {
+        try {
+            retriesCount++;
+            const result = await functionToRetry();
+            const assertion = expect(result);
+            let negationIsApplied = false;
+
+            for (const { isNegate, method, args = [], key } of assertionStack) {
+                if (isNegate) {
+                    negationIsApplied = true;
+                    continue;
+                }
+
+                try {
```

Comment:
@olehrakwix yeah, it would really be great 😅 I was trying to handle such cases by following call of methods:
```ts
method?.apply(isNegationApplied ? assertion.not : assertion, args);
```

but it doesn't work when assertion are applied by iterating through array, probably negation flag isn't stored in `this`, so I decided to implement custom mechanism that apply negation exactly like chai

Reply:
@olehrakwix fixed